core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/FaServices.java

```diff
@@ -1,5653 +1,5658 @@
+    for (FunctionalExchange anExchange : allFunctionalExchanges) {
+      AbstractFunction targetFunction;
+      if (anExchange.getTarget().equals(selectedFunction)
+          || anExchange.getTarget().eContainer().equals(selectedFunction)) {
+        targetFunction = FunctionExt.getRelatedFunction(anExchange.getSource());
+      } else {
+        targetFunction = FunctionExt.getRelatedFunction(anExchange.getTarget());
+      }
+      if (!targetFunction.getAllocationBlocks().isEmpty()) {
+        returnedList.add(anExchange);
+      } else {
+        // it could be allocated by a role (which is not a block)
+        EList<AbstractTrace> incomingTraces = targetFunction.getIncomingTraces();
+        for (AbstractTrace abstractTrace : incomingTraces) {
+          if (abstractTrace instanceof ActivityAllocation) {
+            returnedList.add(anExchange);
+            break;
+          }
+        }
+      }
+    }
+    return returnedList;
+  }
+
+  private DNodeContainer getDisplayedFunctionContainer(EObject function, Collection<DNodeContainer> containers) {
+    EObject currentFunction = function;
+    while (currentFunction instanceof AbstractFunction) {
+      for (DNodeContainer aContainer : containers) {
+        if (aContainer.getTarget().equals(currentFunction)) {
+          return aContainer;
+        }
+      }
+      currentFunction = currentFunction.eContainer();
+    }
+    return null;
+  }
+
+  /**
+   * @param semantic
+   * @param dDiagram
+   * @return
+   */
+  @Deprecated
+  public DiagramElementMapping getMappingABRole(Role semantic, DDiagram diagram) {
+    String mappingName = MappingConstantsHelper.getMappingABRole(diagram);
+    return DiagramServices.getDiagramServices().getContainerMapping(diagram, mappingName);
+
+  }
+
+  public ContainerMapping getMappingABComponent(EObject eObject, DDiagram diagram) {
+    String mappingName = MappingConstantsHelper.getMappingABComponent(eObject, diagram);
+    return DiagramServices.getDiagramServices().getContainerMapping(diagram, mappingName);
+  }
+
+  @Deprecated
+  public AbstractNodeMapping getMappingFunction(AbstractFunction function, DDiagram diagram) {
+    String mappingName = MappingConstantsHelper.getMappingFunction(diagram);
+    return DiagramServices.getDiagramServices().getAbstractNodeMapping(diagram, mappingName);
+  }
+
+  @Deprecated
+  public AbstractNodeMapping getMappingFunctionPort(FunctionPort port, DDiagram diagram) {
+    String mappingName = MappingConstantsHelper.getMappingFunctionPort(diagram);
+    return DiagramServices.getDiagramServices().getBorderedNodeMapping(diagram, mappingName);
+  }
+
+  @Deprecated
+  public EdgeMapping getMappingFunctionalExchange(DDiagram diagram) {
+    String mappingName = MappingConstantsHelper.getMappingFunctionalExchange(diagram);
+    return DiagramServices.getDiagramServices().getEdgeMapping(diagram, mappingName);
+  }
+
+  @Deprecated
+  public NodeMapping getMappingABAbstractFunction(AbstractFunction functi
```

Comment:
Note that you can use .toList() instead of collect(Collectors.toList()) if you don't need to modify it.